documentation/examples/todoist/todoist.ts

```diff
@@ -1,100 +1,404 @@
-// This import statement gives you access to all parts of the Coda Packs SDK.
 import * as coda from "@codahq/packs-sdk";
 
-// This line creates your new Pack.
+
+// Constants.
+
+const ProjectUrlPatterns: RegExp[] = [
+  new RegExp("^https://todoist.com/app/project/([0-9]+)$"),
+  new RegExp("^https://todoist.com/showProject\\?id=([0-9]+)"),
+];
+
+const TaskUrlPatterns: RegExp[] = [
+  new RegExp("^https://todoist.com/app/project/[0-9]+/task/([0-9]+)$"),
+  new RegExp("^https://todoist.com/showTask\\?id=([0-9]+)"),
+];
+
+
+// Pack setup.
+
 export const pack = coda.newPack();
 
-// We use "user authentication" so that users of this Pack will provide their
-// own credentials, as opposed to "system authentication", which the pack author
-// provides a single set of credentials that apply to all requests from this
-// pack. In this case, users must grab the API token from the "Integrations" tab
-// within Todoist's settings
-pack.setUserAuthentication({type: coda.AuthenticationType.HeaderBearerToken});
+pack.addNetworkDomain("todoist.com");
+
+pack.setUserAuthentication({
+  type: coda.AuthenticationType.OAuth2,
+  // Settings pulled from the Todoist OAuth guide.
+  // https://developer.todoist.com/guides/#oauth
+  authorizationUrl: "https://todoist.com/oauth/authorize",
+  tokenUrl: "https://todoist.com/oauth/access_token",
+  scopes: ["data:read_write"],
+
+  // Determines the display name of the connected account.
+  getConnectionName: coda.makeMetadataFormula(async context => {
+    let url = coda.withQueryParams("https://api.todoist.com/sync/v8/sync", {
+      resource_types: JSON.stringify(["user"]),
+    });
+    let response = await context.fetcher.fetch({
+      method: "GET",
+      url: url,
+    });
+    return response.body.user?.full_name;
+  }),
+});
 
-// When using the fetcher, this is the domain of the API that your pack makes
-// fetcher requests to.
-pack.addNetworkDomain("api.todoist.com");
+// Schemas
+
+// A reference to a synced Project. Usually you can use
+// `coda.makeReferenceSchemaFromObjectSchema` to generate these from the primary
+// schema, but that doesn't work in this case since a Project itself can contain
```

Comment:
Ah so these are references to their own table? That's unfortunate we can't demonstrate references more simply, alas, for another example.

Reply:
I think in the narrow samples I cut from this I will omit the same-table references for simplicity, but having worked through this challenge thought it was worth leaving them in, in case other devs run into this need.